File: galite-core/src/main/java/org/kopi/vkopi/lib/ui/swing/form/KeyNavigator.java
/*
 * Copyright (c) 1990-2022 kopiRight Managed Solutions GmbH
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * $Id: KeyNavigator.java 34961 2016-11-04 17:20:49Z hacheni $
 */

package org.kopi.vkopi.lib.ui.swing.form;

import java.awt.event.ActionEvent;

import javax.swing.AbstractAction;
import javax.swing.JComponent;
import javax.swing.text.JTextComponent;

import org.kopi.galite.visual.form.VField;
import org.kopi.galite.visual.util.PrintJob;
import org.kopi.galite.visual.visual.Action;
import org.kopi.galite.visual.visual.PrinterManager;
import org.kopi.galite.visual.visual.VException;
import org.kopi.vkopi.lib.ui.swing.base.Stateful;
import org.kopi.vkopi.lib.ui.swing.visual.Utils;

public class KeyNavigator extends AbstractAction {

    // ----------------------------------------------------------------------
    // CONSTRUCTOR
    // ----------------------------------------------------------------------

    protected KeyNavigator(int keyCode) {
        super("navigation-key");
        this.keyCode = keyCode;
    }

    public void actionPerformed(final ActionEvent e) {
        final Stateful sharedText;
        final DField fieldView;

        try {
            sharedText = (Stateful) ((JTextComponent) e.getSource()).getDocument();
            // !! find fieldView toDo: find more beautyful way
            fieldView = (DField) Utils.getRoot(((JComponent) e.getSource()), DField.class);
        } catch (ClassCastException cce) {
            return;
        }

        if (sharedText == null) {
            return;
        } else {
            final VField field = (VField) sharedText.getModel();
            if (field == null ||
                    field.getForm() == null ||
                    field.getForm().getActiveBlock() == null ||
                    (field.getForm().getActiveBlock().getActiveField() != field &&
                            keyCode != KEY_NEXT_BLOCK)) {
                return;
            } else {
                processKeyCode(fieldView, field, (JTextComponent) e.getSource());
            }
        }
    }

    protected final void processKeyCode(final DField fieldView, final VField field, final JTextComponent sharedText) {
        Action action;

        switch (keyCode) {
            case KEY_NEXT_FIELD:
                action = new Action("keyKEY_TAB") {
                    public void execute() {
                        field.getBlock().getForm().getActiveBlock().gotoNextField();
                    }
                };
                break;
            case KEY_PREV_FIELD:
                action = new Action("keyKEY_STAB") {
                    public void execute() {
                        field.getBlock().getForm().getActiveBlock().gotoPrevField();
                    }
                };
                break;
            case KEY_NEXT_BLOCK:
                action = new Action("keyKEY_BLOCK") {
                    public void execute() {
                        field.getBlock().getForm().gotoNextBlock();
                    }
                };
                break;
            case KEY_REC_UP:
                action = new Action("keyKEY_REC_UP") {
                    public void execute() {
                        fieldView.getBlockView().getModel().gotoPrevRecord();
                    }
                };
                break;
            case KEY_REC_DOWN:
                action = new Action("keyKEY_REC_DOWN") {
                    public void execute() {
                        fieldView.getBlockView().getModel().gotoNextRecord();
                    }
                };
                break;
            case KEY_REC_FIRST:
                action = new Action("keyKEY_REC_FIRST") {
                    public void execute() {
                        field.getBlock().getForm().getActiveBlock().gotoFirstRecord();
                    }
                };
                break;
            case KEY_REC_LAST:
                action = new Action("keyKEY_REC_LAST") {
                    public void execute() {
                        field.getBlock().getForm().getActiveBlock().gotoLastRecord();
                    }
                };
                break;
            case KEY_EMPTY_FIELD:
                action = new Action("keyKEY_ALTENTER") {
                    public void execute() {
                        field.getBlock().getForm().getActiveBlock().gotoNextEmptyMustfill();
                    }
                };
                break;
            case KEY_DIAMETER:
                action = new Action("keyKEY_DIAMETER") {
                    public void execute() {
                        JTextComponent text = sharedText;

                        text.replaceSelection("\u00D8");
                        text.select(text.getSelectionStart() + 1, text.getSelectionStart() + 1);
                    }
                };
                // execute it in the event-dispatching-thread!
                fieldView.getBlockView().getFormView().performBasicAction(action);
                return;
            case KEY_ESCAPE:
                fieldView.getBlockView().getFormView().closeWindow();
                action = null;
                break;
            case KEY_PRINTFORM:
                action = new Action("keyKEY_ALTENTER") {
                    public void execute() {
                        PrintJob job = null;
                        try {
                            job = fieldView.getBlockView().getFormView().printForm();
                        } catch (VException e) {
                            e.printStackTrace();
                        }

                        PrinterManager.Companion.getPrinterManager().getCurrentPrinter().print(job);

                    }
                };
                break;
            default:
                action = processSpecificKeyCode(fieldView, field);
        }
        if (action != null) {
            fieldView.getBlockView().getFormView().performAsyncAction(action);
        }
    }

    /**
     * Subclasses must override this method to process their specific
     * keys they are added by addSpecificNavigationKeys
     */
    protected Action processSpecificKeyCode(final DField fieldView, final VField field) {
        Action action;

        switch (keyCode) {
            case KEY_PREV_VAL:
                action = new Action("keyKEY_LIST_UP") {
                    public void execute() {
                        fieldView.getRowController().previousEntry();
                    }
                };
                break;
            case KEY_NEXT_VAL:
                action = new Action("keyKEY_LIST_DOWN") {
                    public void execute() {
                        fieldView.getRowController().nextEntry();
                    }
                };
                break;
            default:
                action = null;
        }

        return action;
    }

    // ----------------------------------------------------------------------
    // get a navigator
    // ----------------------------------------------------------------------

    public static KeyNavigator getKeyNavigator(int code) {
        return navigators[code];
    }

    // ----------------------------------------------------------------------
    // DATA MEMBERS
    // ----------------------------------------------------------------------

    protected final int keyCode;

    public static final int KEY_NEXT_FIELD = 0;
    public static final int KEY_PREV_FIELD = 1;
    public static final int KEY_REC_UP = 2;
    public static final int KEY_REC_DOWN = 3;
    public static final int KEY_REC_FIRST = 4;
    public static final int KEY_REC_LAST = 5;
    public static final int KEY_EMPTY_FIELD = 6;
    public static final int KEY_NEXT_BLOCK = 7;
    public static final int KEY_PREV_VAL = 8;
    public static final int KEY_NEXT_VAL = 9;
    public static final int KEY_DIAMETER = 10;
    public static final int KEY_ESCAPE = 11;
    public static final int KEY_PRINTFORM = 12;

    public static final KeyNavigator[] navigators;

    static {
        navigators = new KeyNavigator[]{
                new KeyNavigator(KEY_NEXT_FIELD),
                new KeyNavigator(KEY_PREV_FIELD),
                new KeyNavigator(KEY_REC_UP),
                new KeyNavigator(KEY_REC_DOWN),
                new KeyNavigator(KEY_REC_FIRST),
                new KeyNavigator(KEY_REC_LAST),
                new KeyNavigator(KEY_EMPTY_FIELD),
                new KeyNavigator(KEY_NEXT_BLOCK),
                new KeyNavigator(KEY_PREV_VAL),
                new KeyNavigator(KEY_NEXT_VAL),
                new KeyNavigator(KEY_DIAMETER),
                new KeyNavigator(KEY_ESCAPE),
                new KeyNavigator(KEY_PRINTFORM)
        };
    }
}

File: galite-core/src/main/kotlin/org/kopi/galite/ui/vaadin/form/KeyNavigator.kt
/*
 * Copyright (c) 2013-2020 kopiLeft Services SARL, Tunis TN
 * Copyright (c) 1990-2020 kopiRight Managed Solutions GmbH, Wien AT
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
package org.kopi.galite.ui.vaadin.form

import java.io.IOException

import org.kopi.galite.form.VField
import org.kopi.galite.ui.vaadin.event.TextFieldListener
import org.kopi.galite.ui.vaadin.field.InputTextField
import org.kopi.galite.util.PrintException
import org.kopi.galite.visual.Action
import org.kopi.galite.visual.PrinterManager
import org.kopi.galite.visual.VExecFailedException
import org.kopi.galite.visual.VWindow

/**
 * Text field key navigation.
 */
class KeyNavigator(private val model: VField?, private val box: InputTextField<*>?) : TextFieldListener {

  //---------------------------------------------------
  // IMPLEMENTATIONS
  //---------------------------------------------------
  override fun gotoNextField() {
    performAction(object : Action("keyKEY_TAB") {
      override fun execute() {
        if (model != null) {
          model.block!!.form.getActiveBlock()!!.gotoNextField()
        }
      }
    })
  }

  override fun gotoPrevField() {
    performAction(object : Action("keyKEY_STAB") {
      override fun execute() {
        if (model != null) {
          model.block!!.form.getActiveBlock()!!.gotoPrevField()
        }
      }
    })
  }

  override fun gotoNextBlock() {
    performAction(object : Action("keyKEY_BLOCK") {
      override fun execute() {
        if (model != null) {
          model.block!!.form.gotoNextBlock()
        }
      }
    })
  }

  override fun gotoPrevRecord() {
    performAction(object : Action("keyKEY_REC_UP") {
      override fun execute() {
        if (model != null) {
          model.block!!.gotoPrevRecord()
        }
      }
    })
  }

  override fun gotoNextRecord() {
    performAction(object : Action("keyKEY_REC_DOWN") {
      override fun execute() {
        if (model != null) {
          model.block!!.gotoNextRecord()
        }
      }
    })
  }

  override fun gotoFirstRecord() {
    performAction(object : Action("keyKEY_REC_FIRST") {
      override fun execute() {
        if (model != null) {
          model.block!!.form.getActiveBlock()!!.gotoFirstRecord()
        }
      }
    })
  }

  override fun gotoLastRecord() {
    performAction(object : Action("keyKEY_REC_LAST") {
      override fun execute() {
        if (model != null) {
          model.block!!.form.getActiveBlock()!!.gotoLastRecord()
        }
      }
    })
  }

  override fun gotoNextEmptyMustfill() {
    performAction(object : Action("keyKEY_ALTENTER") {
      override fun execute() {
        if (model != null) {
          model.block!!.form.getActiveBlock()!!.gotoNextEmptyMustfill()
        }
      }
    })
  }

  override fun closeWindow() {
    performAction(object : Action("keyKEY_ESCAPE") {
      override fun execute() {
        if (model != null) {
          model.block!!.form.close(VWindow.CDE_QUIT)
        }
      }
    })
  }

  override fun printForm() {
    performAction(object : Action("keyKEY_ALTENTER") {
      override fun execute() {
        if (model != null) {
          try {
            val job = model.getDisplay()!!.getBlockView().getFormView().printForm()
            PrinterManager.getPrinterManager().getCurrentPrinter().print(job!!)
          } catch (e: PrintException) {
            throw VExecFailedException(e.message)
          } catch (e: IOException) {
            throw VExecFailedException(e.message)
          }
        }
      }
    })
  }

  override fun previousEntry() {
    performAction(object : Action("keyKEY_LIST_UP") {
      override fun execute() {
        if (model != null) {
          (model.getDisplay() as DField).rowController.previousEntry()
        }
      }
    })
  }

  override fun nextEntry() {
    performAction(object : Action("keyKEY_LIST_UP") {
      override fun execute() {
        if (model != null) {
          (model.getDisplay() as DField).rowController.nextEntry()
        }
      }
    })
  }

  override fun onQuery(query: String?) {
    model!!.getForm().performAsyncAction(object : Action() {
      override fun execute() {
        val suggestions = model.getSuggestions(query)
        if (box != null && suggestions != null) {
          /*BackgroundThreadHandler.access(Runnable { TODO
            box.setSuggestions(suggestions, query)
          })*/
        }
      }
    })
  }

  /*fun onSuggestion(suggestion: AutocompleteSuggestion?) { TODO
    // not working cause objects passed in the the shared state are not properly
    // serialized. We will set the field value at the client side and fire a go to next field
    // here to convert the field displayed value to model value.
    // model.setObject(model.block.getActiveRecord(), suggestion.getValue());
  }*/

  override fun autofill() {
    (model!!.getDisplay() as DField).performAutoFillAction()
  }

  /**
   * Executes the given action in the event dispatch handler.
   * @param action The action to be executed.
   */
  protected fun performAction(action: Action?) {
    if (action != null && model != null) {
      model.getForm().performAsyncAction(action)
    }
  }
}
